af/shovel/centrifugation.py

```diff
@@ -326,163 +329,491 @@ def __flush_stubborn(self):
-                    use_report = False
-            elif t == '/report':
-                use_report = True
-            elif t == 'datum' and use_report:
-                datum_count += 1
-
+            elif doc['type'] == '/file':
+                files[filename] = b64decode(doc['file_sha1'])
+    return files
+
+def load_autoclaved_db_todo(pgconn, files, bucket):
+    # Returns: {files to ingest}, {files to re-ingest}, {files to re-process}
+    #   ingest    ~ create new measurement records
+    #   reingest  ~ preserve measurement `msm_no`, update binary representation of autoclaved file & re-parse data
+    #   reprocess ~ just re-parse data (or some sub-set of tables)
+    files = files.copy() # shallow copy to pop values from
+    reingest, reprocess = set(), set()
     with pgconn.cursor() as c:
-        c.execute('SELECT file_sha1, filename FROM autoclaved WHERE bucket_date = %s', (bucket,))
-        files_db = {(str(_[0]), _[1]) for _ in c} # NB: file_sha1 is returned as buffer()
-        if files_db == files:
-            # that's just a safety check, these tables are commited within single transaction
-            c.execute('SELECT COUNT(*) FROM autoclaved JOIN report USING (autoclaved_no) WHERE bucket_date = %s', (bucket,))
-            report_db = list(c)[0][0]
-            if report_db != report_count:
-                raise RuntimeError('Unexpected DB mismatch in report count', bucket, report_count, report_db)
-            c.execute('SELECT COUNT(*) FROM autoclaved JOIN report USING (autoclaved_no) JOIN measurement USING (report_no) WHERE bucket_date = %s', (bucket,))
-            datum_db = list(c)[0][0]
-            if datum_db != datum_count:
-                raise RuntimeError('Unexpected DB mismatch in measurement count', bucket, datum_count, datum_db)
-            if len(files) != 0:
-                c.execute('SELECT MIN(code_ver), MAX(code_ver) FROM autoclaved WHERE bucket_date = %s', (bucket,))
-                vermin, vermax = list(c)[0]
-            else:
-                vermin, vermax = CODE_VER, CODE_VER # as DB will return NULL in this case
-            if vermin != vermax:
-                raise RuntimeError('Unexpected DB mismatch in code versions', bucket, vermin, vermax)
-            return vermin
-        elif len(files_db) == 0: # and len(files) > 0 as sets are not equal :)
-            return None
+        c.execute('SELECT filename, file_sha1, code_ver FROM autoclaved WHERE bucket_date = %s', (bucket,))
+        for filename, file_sha1, code_ver in c:
+            file_sha1 = str(file_sha1) # file_sha1 is returned as buffer()
+            if filename not in files:
+                raise RuntimeError('Unknown filename in autoclaved PG table present in autoclaved/index.json.gz', filename)
+            elif files[filename] != file_sha1: # autoclaved file changed, e.g. PII cleanup
+                reingest.add(filename)
+            elif code_ver < CODE_VER: # including CODE_VER_REPROCESS
```

Comment:
I don't understand the comment here. Should this comparison be against `CODE_VER_REPROCESS` or include `CODE_VER_REPROCESS` in some way?

Reply:
Actually, the comment is there to make code grepable for "CODE_VER_REPROCESS" string :)